Expensify/react-native-onyx
# Use `MemoryOnlyProvider` as fallback to allow the app to work without storage

@marcaaron @tgolen 

### Details


This PR adds a `MemoryOnlyProvider` fallback solution and changes the interface of `Storage` (`lib/storage/index.js`) to allow multiple simultaneous storage providers.

This PR also improves the structure of everything in `lib/storage` and establishes a consistent structure for providers and order for all the functions within them.

### Related Issues

https://github.com/Expensify/App/issues/29403

### Automated Tests


This PR changes the way we use the storage but the functionality of the library to the outer world is the same. Therefore no new tests were added.

### Manual Tests


Verify that no flows and functionality were broken by the changes. Check for console errors regarding Onyx.

### Author Checklist


### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



https://github.com/Expensify/react-native-onyx/assets/20173411/fc014b32-4877-443e-88de-b743269c1319

</details>

<details>
<summary>MacOS: Desktop</summary>



</details>
